cluster-autoscaler/proposals/parallel_drain.md

```diff
@@ -0,0 +1,246 @@
+
+The actuation will be done in a similar fashion as in the old algorithm. For a given set of nodes to be scaled down we will synchronously taint the nodes. Then separate goroutines to perform draining and node deletions will be run.
+
+Current scale-down algorithm uses SoftTainting to limit the chance that new pods will be moved toward scaled down candidates. New algorithm will use the same mechanism for nodes in the candidate\_set.
+
+The state in the scale-down algorithm will be updated incrementally based on the changes to the set of nodes and pods in the cluster snapshot.
+
+## Detailed design
+
+### Existing code refactoring
+
+The existing scale down code lives mostly in the ScaleDown object, spanning scale\_down.go file with 1.5k lines of code. As a part of this effort, ScaleDown object will undergo refactoring, extracting utils to a separate file. ScaleDown itself will become an interface with two implementations (both relying on common utils): existing version and the new algorithm described below. This will allow easy switching between algorithms with a flag: different flag values will pick different ScaleDown interface implementations.
+
+As a part of the refactoring, we will combine `FastGetPodsToMove` and `DetailedGetPodsForMove` into a single `PodsToMove` function. The `checkReferences` flag will be dropped and we will always do a detailed check. We are now using listers so doing a detailed check does not add extra API calls and should not add too much to execution time.
+
+### Algorithm State
+
+The algorithm is stateful. On each call to UpdateNodeInfos (happening every CA loop iteration) the state will be updated to match the most recent cluster state snapshot. The most important fields to be held in algorithm state are:
+
+*   deleted\_set: set of names of nodes which are being deleted (implementation-wise we will probably use NodeDeletionTracker)
+*   candidate\_set: set of names for nodes which are can be deleted
+*   non\_candidate\_set: set of names of nodes for which we tried to simulate drain and it failed. For each node we keep time when the simulation was done.
+*   pod\_destination\_hints:
+    *   stores destination nodes computed during draining simulation. For each pod from nodes in candidate\_set and deleted\_set it keeps the destination node assigned during simulation.
+    *   map: source node names -> pod UID -> destination node names
+*   previous\_node\_infos
```

Comment:
It is used to return unneeded nodes for deletion. The candidate\_set is only a list of identifiers (node names).

Reply:
So we only really use it later in the same loop (we may technically keep it to next loop, but only because there is no obvious place to clear it and not much point doing so)? Or do we actually use it in the next loop in any way? I don't particularly like the idea of operating on old and potentially outdated nodeInfos and I don't see anything in the proposed algorithm that would require us to do so.

Reply:
Yeah, it was meant as metadata for subsequent fetching of unneeded nodes / candidate_set. Removed, since it is both not the greatest idea and too low-level to include here.